### PR TITLE
Please add comments or suggestions

### PR DESCRIPTION
**Pull request by russellballestrini**
Please add comments or suggestions.

I think I addressed most of the suggestions in the original PR found here: https://github.com/remind101/dbsnap-verify/pull/1

A version of this code has been running on and off in stage for many months.

### CODE REVIEW

**Comment by russellballestrini** [comment]
@remind101/ops-engineering bump on this.

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+		cp -rf env/lib/python*/site-packages/* ./dist
+		cd ./dist && zip -r "../artifacts/lambda-dbsnap-verify-$(COMMIT_HASH).zip" .
+
+### Python3 ###
+
+test3: clean venv
+		. env/bin/activate && python setup.py pytest
+
+venv:
+		python3 -m venv env && env/bin/pip install setuptools --upgrade
```
Comment (russellballestrini) [inline Makefile]: @phobologic this is what I was talking about during yesterday's stand up.

File: dbsnap_verify/__init__.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    transition_state,
+    get_or_create_state_doc,
+    clean_state_doc,
+)
+
+import boto3
+
+# retry 3 times on errors.
+from botocore.config import Config
+BOTO3_CONFIG = Config(retries={"max_attempts":3})
```
Comment (russellballestrini) [inline dbsnap_verify/__init__.py]: @ejholmes I stole this idea from your stacker PR

File: Makefile (hunk 1 above)
Comment (phobologic) [reply Makefile]: Yeah, the pipenv stuff handles a bit more than this - it's basically looking to be bundler, which is useful (see their use of Pipfile).  This just replaces virtualenv it looks like.

File: dbsnap_verify/__init__.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+
+def modify(state_doc, rds_session):
+    """modify: currently modifying the temporary RDS db instance
+    settings to allow the dbsnap-verify tool to access it."""
+    logger.info(
+        "Modifying %s master password and security groups",
+        state_doc["tmp_database"]
+    )
+    security_group_ids = state_doc["database_security_group_ids"]
+    if isinstance(security_group_ids, basestring):
```
Comment (phobologic) [inline dbsnap_verify/__init__.py]: Seems like this happens often enough that it should probably be a helper function.

Comment (phobologic) [reply dbsnap_verify/__init__.py]: Or maybe the state_doc should handle it automatically?

File: dbsnap_verify/state_doc.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    if state_doc is None:
+        state_doc = create_state_doc(config)
+    return state_doc
+
+
+def get_or_create_state_doc_in_path(config):
+    try:
+        with open(config["state_doc_path"], 'r') as json_file:
+            state_doc = json.load(json_file)
+    except IOError:
```
Comment (phobologic) [inline dbsnap_verify/state_doc.py]: Can't IOError also mean permission issues?  Can probably tell that from the error that will come from trying ot create the doc, but not certain.

File: dbsnap_verify/__main__.py [hunk 5]
```diff
@@ -0,0 +1,16 @@
+import argparse
+
+import json
+
+from . import handler
+
+def main():
+    parser = argparse.ArgumentParser(description = 'verify AWS RDS DB snapshots.')
+    parser.add_argument('config', help='The path to JSON config.')
```
[the remark this answers is not in the document]
Comment (russellballestrini) [reply dbsnap_verify/__main__.py]: Neat I didn't know about that. That said, it's not clear how to interact with this with a context manager (`with`) so I guess cleaning up the open file handle needs to be don't manually?

File: dbsnap_verify/state_doc.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+    state_doc_json = json.dumps(state_doc, indent=2)
+    s3 = boto3.client("s3")
+    s3.put_object(
+        Bucket=get_state_doc_bucket(state_doc),
+        Key=state_doc_s3_key(state_doc["database"]),
+        Body=state_doc_json,
+    )
+
+
+def set_state_doc_in_s3(state_doc):
```
[the remark this answers is not in the document]
Comment (russellballestrini) [reply dbsnap_verify/state_doc.py]: fixed

File: dbsnap_verify/state_doc.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+
+import time
+
+import boto3
+
+from .rds_funcs import dbsnap_verify_db_id
+
+DB_ID_PREFIX_LEN = 14
+
+# TODO: it feels like `state_doc` should be a dict-like class
```
[the remark this answers is not in the document]
Comment (russellballestrini) [reply dbsnap_verify/state_doc.py]: Agreed, please review #3

File: dbsnap_verify/__init__.py (hunk 3 above)
Comment (russellballestrini) [reply dbsnap_verify/__init__.py]: Fixed see PR #3

File: dbsnap_verify/__init__.py [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+    """wait: currently waiting for the next snapshot to appear."""
+    verified_snapshot_id = state_doc.get("snapshot_verified", "any snapshot")
+    logger.info(
+        "Looking for a snapshot of %s (newer then %s)",
+        state_doc["database"],
+        verified_snapshot_id,
+    )
+    snapshot_desc = get_latest_snapshot(rds_session, state_doc["database"])
+    if snapshot_desc["DBSnapshotIdentifier"] != verified_snapshot_id:
+        # if the latest snapshot is older then the minimum, restore it.
```
[the remark this answers is not in the document]
Comment (russellballestrini) [reply dbsnap_verify/__init__.py]: Fixed see PR #3

File: dbsnap_verify/__init__.py [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+    basestring
+except NameError:
+    basestring = str
+
+
+def wait(state_doc, rds_session):
+    """wait: currently waiting for the next snapshot to appear."""
+    verified_snapshot_id = state_doc.get("snapshot_verified", "any snapshot")
+    logger.info(
+        "Looking for a snapshot of %s (newer then %s)",
```
[the remark this answers is not in the document]
Comment (russellballestrini) [reply dbsnap_verify/__init__.py]: Fixed see PR #3

**Comment by russellballestrini** [comment]
Hey @phobologic a lot has changed since your last review. Could I get another before I merge?

**Comment by russellballestrini** [comment]
@ejholmes please take another pass when you have a chance.

**Comment by russellballestrini** [comment]
@Lowercases feel free to add comments if you want to.

File: dbsnap_verify/state_doc.py [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+            json_file.write(self.to_json)
+
+    def _load_state_doc_from_path(self):
+        """Returns a JSON String State Document."""
+        with open(self.state_doc_file_path, 'r') as json_file:
+            return json_file.read()
+
+    def transition_state(self, new_state):
+        """Change the state of the StateDoc and save to persistence."""
+        self.states.append(
```
Comment (ejholmes) [inline dbsnap_verify/state_doc.py]: There's a state machine, but I don't see anywhere where allowed state transitions are defined and validated. (e.g. what are valid states to transition to from`wait`?).

You have a pretty good diagram of the state transitions, but I don't see where that's verified in code (I could be missing it).

File: README.rst [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+
+You may build an AWS Lambda zip. This command assumes you have Python ``virtualenv`` installed::
+
+ make build-lambda
+
+In our case use use a Cloudwatch Rule Event Trigger to invoke our Lambda at a ``rate(15 minutes)``.
+
+The payload of this Cloudwatch Rule is a Static JSON value and is in the same form as the config used for the CLI.
+
+An example `dbsnap-verify-config.json <https://github.com/remind101/dbsnap-verify/blob/import/tests/fixtures/config_or_event.json>`_ may be found here.
```
Comment (ejholmes) [inline README.rst]: This link is broken.

File: README.rst [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+
+An `example_state_doc.json <https://github.com/remind101/dbsnap-verify/blob/import/tests/fixtures/example_state_doc.json>`_ may be found here.
+
+
+State Machine Diagram
+====================================
+
+Here is a diagram of the state machine transitions and states.
+
+.. image:: https://github.com/remind101/dbsnap-verify/raw/import/dbsnap-verify-rds-snapshot-verification-lambda-s3-state-machine.png
```
Comment (ejholmes) [inline README.rst]: This image is broken.

File: dbsnap_verify/rds_funcs.py [hunk 13]
```diff
@@ -0,0 +1,267 @@
...
+            connection where the database is located.
+        db_id (str): The RDS database instance identifier to destroy.
+    """
+    if db_arn is None:
+        description = get_database_description(session, db_id)
+        db_arn = description["DBInstanceArn"]
+
+    tags = get_tags_for_rds_arn(session, db_arn)
+
+    if tags.get("dbsnap-verify", "false") != "true":
```
Comment (Lowercases) [inline dbsnap_verify/rds_funcs.py]: This is a good idea.

This is one of the instances where I'd go for a global constant, TAG_NAME and TAG_VALUE for "dbsnap-verify" and "true" (or w/e) -- this code is critical so I think it's best to fool-proff it against future edits.

**Comment by Lowercases** [comment]
It looks good in general! As for the code itself, I think it goes over my head so I'll defer to trust it if it's been throughly tested :) I did leave a single suggestion nevertheless.

File: dbsnap_verify/state_doc.py (hunk 10 above)
Comment (russellballestrini) [reply dbsnap_verify/state_doc.py]: Please see https://github.com/remind101/dbsnap-verify/pull/2/commits/32b626b7622c2add0e266ab0bba7a429c1f7ac7e

**Comment by russellballestrini** [comment]
@remind101/ops-engineering please have another pass before I merge.

**Comment by ejholmes** [comment]
Other than what we talked about with alarms, this lgtm! :+1:

**Pull request merged**